src/sources/docker.rs

```diff
@@ -379,256 +462,121 @@ impl DockerSource {
-            .inspect()
-            .map_err(|error| error!(message="Fetching container details failed",%error))
-            .and_then(|details| {
-                ContainerMetadata::from_details(&details)
-                    .map_err(|error| error!(message="Metadata extraction failed",%error))
-            });
-
         let this = self.clone();
-        let task = metadata_fetch.and_then(move |metadata| {
-            this.start_event_stream(ContainerLogInfo::new(id, metadata, this.core.now_timestamp))
+        tokio::spawn(async move {
+            Arc::clone(&this.core)
+                .docker
+                .inspect_container(id.clone().as_str(), None::<InspectContainerOptions>)
+                .map_err(|error| error!(message = "Fetching container details failed", %error))
+                .and_then(|details| async move {
+                    ContainerMetadata::from_details(details)
+                        .map_err(|error| error!(message = "Metadata extraction failed", %error))
+                })
+                .and_then(|metadata| async move {
+                    let info = ContainerLogInfo::new(id, metadata, this.core.now_timestamp);
+                    this.start_event_stream(info).await;
+                    Ok::<(), ()>(())
+                })
+                .await
         });
 
-        tokio01::spawn(task);
-
         ContainerState::new()
     }
 
     /// If info is present, restarts event stream which will run until shutdown.
     fn restart(&self, container: &mut ContainerState) {
         if let Some(info) = container.take_info() {
-            tokio01::spawn(self.start_event_stream(info));
+            let this = self.clone();
+            tokio::spawn(async move { this.start_event_stream(info).await });
         }
     }
 
-    fn start_event_stream(&self, info: ContainerLogInfo) -> impl Future<Item = (), Error = ()> {
+    async fn start_event_stream(&self, info: ContainerLogInfo) {
         // Establish connection
-        let mut options = LogsOptions::builder();
-        options
-            .follow(true)
-            .stdout(true)
-            .stderr(true)
-            .since(info.log_since())
-            .timestamps(true);
-
-        let mut stream = self
-            .core
-            .docker
-            .containers()
-            .get(info.id.as_str())
-            .logs(&options.build());
+        let options = Some(LogsOptions {
+            follow: true,
+            stdout: true,
+            stderr: true,
+            since: info.log_since(),
+            timestamps: true,
+            ..Default::default()
+        });
+
+        let stream = self.core.docker.logs(info.id.as_str(), options);
         info!(
             message = "Started listening logs on docker container",
             id = field::display(info.id.as_str())
         );
 
         // Create event streamer
-        let mut state = Some((self.main_send.clone(), info));
+        let mut info = Some(info);
```

Comment:
We can just use info directly without the option thanks to wonderful `async`.

Reply:
Indeed, with current code we can, thanks! I was need `Option<T>` for some code in middle of changes.